File: runtime/flutter/material/materialPageRoute.ts
import { Type } from "../../dart/core/type";
import { RuntimeBaseClass } from "../../runtimeBaseClass";
import { JITAllocatingRTManagedBox } from "../../syntheticBox";
import { BuildContext } from "../buildContext";
import { Widget } from "../widget";

export interface MaterialPageRouteProps {
    builder: (context: BuildContext) => Widget;
    maintainState?: boolean | undefined;
    fullscreenDialog?: boolean | undefined;
}

declare const flutter: {
    material: {
        materialPageRoute: (
            this: void,
            props: MaterialPageRouteProps
        ) => MaterialPageRoute;
    };
};

export class MaterialPageRoute
    extends JITAllocatingRTManagedBox<MaterialPageRouteProps, MaterialPageRoute>
    implements RuntimeBaseClass {
    public readonly internalRuntimeType = new Type(MaterialPageRoute);
    public props: MaterialPageRouteProps;
    public constructor(props: MaterialPageRouteProps) {
        super();

        this.props = props;

        if (this.props.maintainState === undefined) {
            this.props.maintainState = true;
        }

        if (this.props.fullscreenDialog === undefined) {
            this.props.fullscreenDialog = false;
        }
    }

    public unwrap(): MaterialPageRoute {
        return flutter.material.materialPageRoute(this.props);
    }
}

File: lib/builtins/flutter/material/materialPageRoute.dart
import 'package:flua/hydroState.dart';
import 'package:flua/vm/closure.dart';
import 'package:flua/vm/context.dart';
import 'package:flua/builtins/flutter/syntheticBox.dart';
import 'package:flua/vm/table.dart';
import 'package:flutter/material.dart';

loadMaterialPageRoute(
    {@required HydroState luaState, @required HydroTable table}) {
  table["materialPageRoute"] = makeLuaDartFunc(func: (List<dynamic> args) {
    return [
      MaterialPageRoute(
          builder: (BuildContext context) {
            Closure closure = args[0]["builder"];
            var res =
                closure.dispatch([null, context], parentState: luaState)[0];
            return maybeUnwrapAndBuildArgument<Widget>(res,
                parentState: luaState);
          },
          maintainState: args[0]["maintainState"],
          fullscreenDialog: args[0]["fullscreenDialog"])
    ];
  });
}
